pinot-controller/src/main/java/org/apache/pinot/controller/cursors/ResponseStoreCleaner.java

```diff
@@ -0,0 +1,222 @@
+      Map<String, String> requestHeaders)
+      throws Exception {
+    String protocol = _controllerConf.getControllerBrokerProtocol();
+    int portOverride = _controllerConf.getControllerBrokerPortOverride();
+    List<String> brokerUrls = new ArrayList<>();
+    for (InstanceInfo broker : brokers.values()) {
+      int port = portOverride > 0 ? portOverride : broker.getPort();
+      brokerUrls.add(String.format(QUERY_RESULT_STORE, protocol, broker.getHost(), port));
+    }
+    LOGGER.debug("Getting running queries via broker urls: {}", brokerUrls);
+    Map<String, String> strResponseMap = getResponseMap(requestHeaders, brokerUrls, "GET", HttpGet::new);
+    return strResponseMap.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> {
+      try {
+        return JsonUtils.stringToObject(e.getValue(), new TypeReference<>() {
+        });
+      } catch (IOException ex) {
+        throw new RuntimeException(ex);
+      }
+    }));
+  }
+
+  private <T extends HttpUriRequestBase> Map<String, String> getResponseMap(Map<String, String> requestHeaders,
+      List<String> brokerUrls, String methodName, Function<String, T> httpRequestBaseSupplier)
+      throws Exception {
+    List<Pair<String, String>> urlsAndRequestBodies = new ArrayList<>(brokerUrls.size());
+    brokerUrls.forEach((url) -> urlsAndRequestBodies.add(Pair.of(url, "")));
+
+    CompletionService<MultiHttpRequestResponse> completionService =
+        new MultiHttpRequest(_executor, _connectionManager).execute(urlsAndRequestBodies, requestHeaders,
+            ResponseStoreCleaner.TIMEOUT_MS, methodName, httpRequestBaseSupplier);
+    Map<String, String> responseMap = new HashMap<>();
+    List<String> errMessages = new ArrayList<>(brokerUrls.size());
+    for (int i = 0; i < brokerUrls.size(); i++) {
+      try (MultiHttpRequestResponse httpRequestResponse = completionService.take().get()) {
+        // The completion order is different from brokerUrls, thus use uri in the response.
+        URI uri = httpRequestResponse.getURI();
+        int status = httpRequestResponse.getResponse().getCode();
+        String responseString = EntityUtils.toString(httpRequestResponse.getResponse().getEntity());
+        // Unexpected server responses are collected and returned as exception.
+        if (status != 200) {
+          throw new Exception(
+              String.format("Unexpected status=%d and response='%s' from uri='%s'", status, responseString, uri));
+        }
+        responseMap.put((getInstanceKey(uri.getHost(), Integer.toString(uri.getPort()))), responseString);
+      } catch (Exception e) {
+        LOGGER.error("Failed to get queries", e);
+        // Can't just throw exception from here as there is a need to release the other connections.
+        // So just collect the error msg to throw them together after the for-loop.
+        errMessages.add(e.getMessage());
```

Comment:
We'll lose the stack trace context for other exceptions (apart from the one we're throwing ourselves above) that are caught and handled here.

Reply:
Yes. I had to work with errors with a client due to TLS issues. The error message was informative with URL and message in it. Stack trace is typically of a network library and I didnt require it.